### PR TITLE
Use if let instead of match when only matching a single variant (clippy::single_match)

### PR DESCRIPTION
**Pull request by matthiaskrgr**
Makes code more compact and reduces nesting.

### CODE REVIEW

File: src/librustc/hir/map/mod.rs [hunk 1]
```diff
@@ -713,14 +713,17 @@ impl<'hir> Map<'hir> {
...
                 }
-                Node::Expr(expr) => match expr.kind {
-                    ExprKind::Match(_, _, _) => return Some(expr),
-                    _ => {}
-                },
-                Node::Stmt(stmt) => match stmt.kind {
-                    StmtKind::Local(_) => break,
-                    _ => {}
-                },
+                Node::Expr(expr) => {
```
Comment (Xanewok) [inline src/librustc/hir/map/mod.rs]: drive-by comment: Maybe this'd be better if we matched directly on
```rust
Node::Expr(expr @ ExprKind::Match(...)) => return Some(expr),
```
and similarly below?

Comment (matthiaskrgr) [reply src/librustc/hir/map/mod.rs]: Hm, this won't compile
```
error[E0308]: mismatched types
   --> src/librustc/hir/map/mod.rs:716:35
    |
712 |             match node {
    |                   ---- this expression has type `rustc_hir::hir::Node<'_>`
...
716 |                 Node::Expr(expr @ ExprKind::Match(..)) => {return Some(expr)},
    |                                   ^^^^^^^^^^^^^^^^^^^ expected struct `rustc_hir::hir::Expr`, found enum `rustc_hir::hir::ExprKind`
```
I'm not very familiar with this kind of syntax myself :(

Comment (Xanewok) [reply src/librustc/hir/map/mod.rs]: Oh, whoops, that’s matching on the field `kind`. Sorry for the trouble 🙇

**Comment by estebank** [comment]
The code changes look ok, please deal with the merge conflicts.

**Comment by matthiaskrgr** [comment]
rebased

File: src/librustc_errors/emitter.rs [hunk 2]
```diff
@@ -1719,39 +1719,37 @@ impl EmitterWriter {
...
                                 true,
                             ) {
-                                Err(e) => panic!("failed to emit error: {}", e),
-                                _ => (),
-                            }
+                                panic!("failed to emit error: {}", e);
+                            };
                         } else {
-                            match self.emit_suggestion_default(sugg, &Level::Help, max_line_num_len)
+                            if let Err(e) =
```
Comment (Centril) [inline src/librustc_errors/emitter.rs]: Join these?

File: src/librustc_mir/transform/check_unsafety.rs [hunk 3]
```diff
@@ -396,43 +396,41 @@ impl<'a, 'tcx> UnsafetyChecker<'a, 'tcx> {
...
+                            } else {
+                                continue;
+                            };
+                            self.require_unsafe(
+                                description,
+                                details,
+                                UnsafetyViolationKind::GeneralAndConstFn,
+                            );
+                        }
+                    };
```
Comment (Centril) [inline src/librustc_mir/transform/check_unsafety.rs]: Why is a `;` being inserted here all the time? (Please check those additions in the PR)
